cmd/dockerd: gracefully shut down the API server

corhere opened the pull request:
- Relates to #45057 

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Gave in-flight API requests up to five seconds to complete after all containers have been shut down before the daemon process exits.

**- How I did it**
By leveraging bleeding-edge APIs...from Go 1.8.

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
cmd/dockerd: ignore SIGPIPE using signal.Ignore

The fix to ignore SIGPIPE signals was originally added in the Go 1.4
era. signal.Ignore was first added in Go 1.5.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
cmd/dockerd/trap: don't force exit after cleanup

Always calling os.Exit() on clean shutdown may not always be desirable
as deferred functions are not run. Let the cleanup callback decide
whether or not to call os.Exit() itself. Allow the process to exit the
normal way, by returning from func main().

Simplify the trap.Trap implementation. The signal notifications are
buffered in a channel so there is little need to spawn a new goroutine
for each received signal. With all signals being handled in the same
goroutine, there are no longer any concurrency concerns around the
interrupt counter.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
cmd/dockerd: gracefully shut down the API server

As of Go 1.8, "net/http".Server provides facilities to close all
listeners, making the same facilities in server.Server redundant.
http.Server also improves upon server.Server by additionally providing a
facility to also wait for outstanding requests to complete after closing
all listeners. Leverage those facilities to give in-flight requests up
to five seconds to finish up after all containers have been shut down.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -243,10 +256,30 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
+		apiWG.Add(1)
+		go func(ls net.Listener) {
+			defer apiWG.Done()
+			logrus.Infof("API listen on %s", ls.Addr())
```
I guess we're not using `WithField("listener", ls.Addr())` here to keep the message the same as before. (users _may_ be grepping for this). We could (separate from this PR) decide to change that, and use structured logging.

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -647,17 +675,20 @@ func checkTLSAuthOK(c *config.Config) bool {
-func loadListeners(cli *DaemonCli, tlsConfig *tls.Config) ([]string, error) {
-	if len(cli.Config.Hosts) == 0 {
-		return nil, errors.New("no hosts configured")
+func loadListeners(cfg *config.Config, tlsConfig *tls.Config) ([]net.Listener, []string, error) {
```
❤️ we really need to make a pass at some of these. Too many places where we just pass everything (mostly out of convenience)

thaJeztah reviewed (commented):
love it!! overall LGTM. left some comments / thoughts / suggestions; all relatively minor, but perhaps there's some you think make sense to include

corhere commented on cmd/dockerd/trap/trap.go:
```diff
@@ -4,47 +4,38 @@ import (
 }) {
-	c := make(chan os.Signal, 1)
-	// we will handle INT, TERM here
+	c := make(chan os.Signal, 10)
```
The exact value is rather arbitrary, though I'll add a comment explaining why I chose a value > 1.

corhere commented on cmd/dockerd/trap/trap.go:
```diff
@@ -4,47 +4,38 @@ import (
-	c := make(chan os.Signal, 1)
-	// we will handle INT, TERM here
+	c := make(chan os.Signal, 10)
 	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```
`signal.Notify(c, os.Interrupt)` also has special meaning on Windows. https://pkg.go.dev/os/signal?GOOS=windows#hdr-Windows

thaJeztah replied:
Yes, I was expecting to see windows having its own value, but it looks like it does _not_; Windows uses the POSIX variant; https://github.com/golang/go/blob/3f747d09e441999815fbb0a6025ade0f528967a8/src/os/exec_posix.go#L5
so on Windows it's also an alias for `syscall.SIGINT`

thaJeztah replied:
Oh, wait, it's not an alias...

cpuguy83 commented on cmd/dockerd/trap/trap.go:
```diff
@@ -4,54 +4,38 @@ import (
-	// we will handle INT, TERM, SIGPIPE here
-	signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGPIPE)
+	c := make(chan os.Signal, 10)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```
Maybe `signal.NotifyContext` and we can pass around the context?

corhere replied:
`signal.Notify` would still be needed to implement the "force-quit on three ^C" behaviour.

corhere committed:
cmd/dockerd/trap: log to logrus directly

Logging through a dependency-injected interface value was a vestige of
when Trap was in pkg/signal to avoid importing logrus in a reusable
package: cc4da8112814cdbb00dbf23370f9ed764383de1f.
Now that Trap lives under cmd/dockerd, nobody will be importing this so
we no longer need to worry about minimizing the package's dependencies.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah reviewed (approved):
LGTM
